src/video/wayland/SDL_waylandwindow.c

```diff
@@ -448,8 +489,105 @@ static const struct xdg_toplevel_listener toplevel_listener_xdg = {
     handle_close_xdg_toplevel
 };
 
+#ifdef HAVE_LIBDECOR_H
+static void
+decoration_frame_configure(struct libdecor_frame *frame,
+                           struct libdecor_configuration *configuration,
+                           void *user_data)
+{
+    SDL_WindowData *wind = user_data;
+    SDL_Window *window = wind->sdlwindow;
+    int width, height;
+    enum libdecor_window_state window_state;
+    struct libdecor_state *state;
+
+    /* window size */
+    if (!libdecor_configuration_get_content_size(configuration, frame, &width, &height)) {
+        width = window->w;
+        height = window->h;
+    }
+
+    wind->resize.width = width;
+    wind->resize.height = height;
+
+    wind->resize.pending = SDL_TRUE;
+    wind->resize.configure = SDL_TRUE;
+    Wayland_HandlePendingResize(window);
+    wind->shell_surface.libdecor.initial_configure_seen = SDL_TRUE;
+
+    window->w = wind->resize.width;
+    window->h = wind->resize.height;
+
+    /* window state */
+    if (!libdecor_configuration_get_window_state(configuration, &window_state)) {
+        window_state = LIBDECOR_WINDOW_STATE_NONE;
+    }
+
+    if (window_state & LIBDECOR_WINDOW_STATE_MAXIMIZED) {
+        window->flags |= SDL_WINDOW_MAXIMIZED;
+    }
+    else {
+        window->flags &= ~SDL_WINDOW_MAXIMIZED;
+    }
+
+    if (window_state & LIBDECOR_WINDOW_STATE_ACTIVE) {
+        window->flags |= SDL_WINDOW_INPUT_FOCUS;
+    }
+    else {
+        window->flags &= ~SDL_WINDOW_INPUT_FOCUS;
+    }
+
+    /* The fullscreen flag is already set my some other entity when 'SDL_SetWindowFullscreen'
+     * is called, but we will set it here again in case the compositor requests fullscreen.
+     */
+    if (window_state & LIBDECOR_WINDOW_STATE_FULLSCREEN) {
```

Comment:
This should probably replicate what we do in xdg-shell, but tbh I don't fully recall why the xdg-shell fullscreen behavior is written the way it is... may need to look at the revision history for that.

Reply:
You mean calling `SetFullscreen` in `handle_configure_xdg_toplevel`? That should not be required. The window flags are already set and `libdecor_frame_commit` will take care of setting the right decoration state. The rest is then just that the client redraws the toplevel content at the new fullscreen dimension.

If you see issues with setting the fullscreen mode, let me know.

Reply:
Will do - I should probably git blame to see why the other backends do the toggle, I'm still not sure _how_ it would desync like that and we have other workarounds that we're planning to remove soon, so it could be a bogus fix like the others too.

Reply:
Found the commit: https://github.com/libsdl-org/SDL/commit/19236b462a8af09670c9fd5490cea293d3fc2b75

So apparently it's to accommodate minimize on focus loss? I think?

Reply:
I did it to correctly return to full-screen after the user unminimizes the application (on Mutter, at least). IIRC, without that, it will be restored to windowed mode rather than returning to full-screen mode.

Reply:
Ah, this is a good point - we should check this again with SDL_VIDEO_MINIMIZE_ON_LOST_FOCUS=1 with both xdg-shell and libdecor. If it's no longer necessary it can go in the 2.0.18 removal pile.

Reply:
In my case, I called `SDL_MinimizeWindow()` when a certain key combo is pressed. It wasn't SDL_VIDEO_MINIMIZE_ON_LOST_FOCUS=1 that was minimizing on my behalf.

Reply:
I am totally lost here. We already do a `SetFullscreen(window, driverdata->output);` in the else block here. Could you summarise what is missing here and how this shows up as a bug?

Reply:
Just realized:

```
        window->flags &= ~SDL_WINDOW_FULLSCREEN;
         if (window->flags & SDL_WINDOW_FULLSCREEN) {
```

This will never trip, right? The first line needs to be cut, other than that I think we're all set.

Reply:
This is fixed now.